script/split_tests.py

```diff
@@ -19,88 +19,167 @@ def __init__(
         part.add_to_bucket()
-        self.total_tests += part.total_tests
+        self.approx_execution_time += part.approx_execution_time
+        self.not_measured_files += part.not_measured_files
         self._paths.append(str(part.path))
 
     def get_paths_line(self) -> str:
         """Return paths."""
         return " ".join(self._paths) + "\n"
 
 
+def add_not_measured_files(
+    test: TestFolder | TestFile, not_measured_files: set[TestFile]
+) -> None:
+    """Add not measured files to test folder."""
+    if test.not_measured_files > 0:
+        if isinstance(test, TestFolder):
+            for child in test.children.values():
+                add_not_measured_files(child, not_measured_files)
+        else:
+            not_measured_files.add(test)
+
+
+def sort_by_not_measured(bucket: Bucket) -> tuple[int, float]:
+    """Sort by not measured files."""
+    return (bucket.not_measured_files, bucket.approx_execution_time.total_seconds())
+
+
+def sort_by_execution_time(bucket: Bucket) -> tuple[float, int]:
+    """Sort by execution time."""
+    return (bucket.approx_execution_time.total_seconds(), bucket.not_measured_files)
+
+
 class BucketHolder:
     """Class to hold buckets."""
 
-    def __init__(self, tests_per_bucket: int, bucket_count: int) -> None:
+    def __init__(self, bucket_count: int) -> None:
         """Initialize bucket holder."""
-        self._tests_per_bucket = tests_per_bucket
         self._bucket_count = bucket_count
         self._buckets: list[Bucket] = [Bucket() for _ in range(bucket_count)]
 
     def split_tests(self, test_folder: TestFolder) -> None:
         """Split tests into buckets."""
-        digits = len(str(test_folder.total_tests))
+        avg_execution_time = test_folder.approx_execution_time / self._bucket_count
+        avg_not_measured_files = test_folder.not_measured_files / self._bucket_count
         sorted_tests = sorted(
-            test_folder.get_all_flatten(), reverse=True, key=lambda x: x.total_tests
+            test_folder.get_all_flatten(),
+            key=lambda x: (
+                -x.approx_execution_time,
+                -x.count_children() if isinstance(x, TestFolder) else 0,
+                x.not_measured_files,
+            ),
         )
+        not_measured_tests = set()
         for tests in sorted_tests:
             if tests.added_to_bucket:
                 # Already added to bucket
                 continue
 
-            print(f"{tests.total_tests:>{digits}} tests in {tests.path}")
-            smallest_bucket = min(self._buckets, key=lambda x: x.total_tests)
+            print(f"~{tests.approx_execution_time} execution time for {tests.path}")
             is_file = isinstance(tests, TestFile)
+
+            sort_key = sort_by_execution_time
+            if tests.not_measured_files and tests.approx_execution_time == 0:
+                # If tests are not measured, sort by not measured files
+                sort_key = sort_by_not_measured
```

Comment:
Sorting not measured files separately feels like it overcomplicates the code a lot. The sorting doesn't need to be perfect. In the end it's just a matter if one job finishes 2 or 3 minutes earlier than another.

A good approximation here should be fine. Maybe choose `1 [s]` or `0.5` per test case. That's probably on the high sight but that would only cause tests without time to be distributed more cautiously.